### PR TITLE
refactor(tooltip): refactoring to the new structure + tests

### PR DESCRIPTION
**Pull request by hopkins-tk**
- tooltip refactored to the new style
- tests added

### CODE REVIEW

File: packages/tooltip/Tooltip.test.tsx [hunk 1]
```diff
@@ -1,17 +1,49 @@
...
 
   it( 'should render', () => {
 
-    console.warn('Missing tests for Tooltip!');
-
     return mount(
       <Tooltip label="hello"/>
-    ).wait();
+    ).wait((element) => {
+      expect( element.has('.c-tooltip') ).to.be.true;
```
Comment (Hotell) [inline packages/tooltip/Tooltip.test.tsx]: we use just `to.equal` or `to.deep.equal` for assertions

File: packages/tooltip/Tooltip.tsx [hunk 2]
```diff
@@ -9,34 +9,43 @@ const TooltipTypes = {
...
+type TooltipProps = {
+  type?: TooltipTypesType,
   label: string,
-}
+};
+
+declare global {
+  namespace JSX {
+    interface IntrinsicElements {
+      'bl-tooltip': TooltipProps & Partial<HTMLElement>,
```
Comment (Hotell) [inline packages/tooltip/Tooltip.tsx:21]: root node is `span` so `HTMLSpanElement`

File: packages/tooltip/Tooltip.test.tsx [hunk 3]
```diff
@@ -1,17 +1,49 @@
...
+      <bl-tooltip label="hello"></bl-tooltip>
+    ).wait((element) => {
+      expect( element.has('.c-tooltip') ).to.be.true;
+      expect( element.has('.c-tooltip--right') ).to.be.true;
+      expect( element.one('span').node.getAttribute('aria-label') ).to.equals( 'hello' );
+    });
+
+  });
+
+  it( 'should render with top tooltip', () => {
```
Comment (Hotell) [inline packages/tooltip/Tooltip.test.tsx]: this and following test are doing the same assertions. Unit test should be clean as our codebase, so extract assertion to one function helper.

also the description of `it` isn't really telling anything.

We should probably structure every test to 2 sub `describe` one for `JSX` and one for "raw" `html` custom element usage

**Review by Hotell — CHANGES_REQUESTED** [review]

File: packages/tooltip/Tooltip.test.tsx (hunk 3 above)
Comment (hopkins-tk) [reply packages/tooltip/Tooltip.test.tsx]: Good idea with the two `describe` - I was thinking how we will want to structure this.

File: packages/tooltip/Tooltip.test.tsx (hunk 1 above)
Comment (hopkins-tk) [reply packages/tooltip/Tooltip.test.tsx]: Great :) we have been discussing this yesterday with @elmariofredo  & @davidrus so we can take the opportunity and agree what should be the preferred version.

File: packages/tooltip/Tooltip.tsx (hunk 2 above)
Comment (hopkins-tk) [reply packages/tooltip/Tooltip.tsx:21]: So any other component, that has DIV as a root element should have `HTMLDivElement`?
- is this really correct?
- if we don't pass any attributes to the root element

Comment (Hotell) [reply packages/tooltip/Tooltip.tsx:21]: ok it was late and I was tired :D so nevermind. Let's stick with `HTMLElement` for all intrinsic elements

File: packages/tooltip/Tooltip.test.tsx [hunk 4]
```diff
@@ -1,18 +1,62 @@
-import { Tooltip } from './Tooltip';
-
+import { Tooltip } from './index';
 import { h, mount } from 'bore';
-import { emit } from 'skatejs';
+import {WrappedNode} from "bore";
```
Comment (Hotell) [inline packages/tooltip/Tooltip.test.tsx]: pls merge with `bore` imports Webstorm has it and VSCode as well

File: packages/tooltip/Tooltip.test.tsx [hunk 5]
```diff
@@ -1,18 +1,62 @@
...
+      return mount(
+        <bl-tooltip label="hello"></bl-tooltip>
+      ).wait(expectDefaultRender);
+
+    });
+
+    it( 'should render with top tooltip', () => {
+
+      return mount(
+        <bl-tooltip label="hello" type="top"></bl-tooltip>
```
Comment (Hotell) [inline packages/tooltip/Tooltip.test.tsx]: just a reminder how JSX works ( with bore `h` ):
` <Tooltip label="hello" type="top" />` transforms to:
- `h( Tooltip, {label:'hello', type:'top'} )` 
- which is created by `new Tooltip()`

`<bl-tooltip label="hello" type="top"></bl-tooltip>` transforms to:
- `h('bl-tooltip',{label:'hello', type:'top'})`
- which is created by `document.createElement`

**Review by Hotell — CHANGES_REQUESTED** [review]

File: packages/tooltip/Tooltip.tsx [hunk 6]
```diff
@@ -9,34 +9,47 @@ const TooltipTypes = {
   bottom: 'bottom',
 };
 
-type TooltipTypesType = typeof TooltipTypes;
+type TooltipTypesType = keyof typeof TooltipTypes;
```
Comment (Hotell) [inline packages/tooltip/Tooltip.tsx]: we dont need that suffix TooltipTypes`Type` in name. We now that it is a type alias already ;)

**Review by Hotell — CHANGES_REQUESTED** [review]
just minor tweaks. Otherwise LGMT. and lets wait for new bore release. Good job @hopkins-tk !

**Comment by hopkins-tk** [comment]
This is failing because of skatejs/bore#11

**Comment by Hotell** [comment]
We should temporarely monkey patch bore so we don't get blocked by it

**Comment by hopkins-tk** [comment]
Rebased & refactored to the newest version of tests we have agreed on.

File: packages/tooltip/Tooltip.test.tsx [hunk 7]
```diff
@@ -1,16 +1,99 @@
...
+          <bl-tooltip label="hello <&> Hi" />
+        ).wait(( element ) => {
+
+          expect( element.one('span').node.getAttribute('aria-label') ).toBe( 'hello <&> Hi' );
+
+        });
+
+      });
+
+      it( `should render with correct label using property`, () => {
```
Comment (Hotell) [inline packages/tooltip/Tooltip.test.tsx]: this test is doing exactly the same as last one

Comment (Hotell) [reply packages/tooltip/Tooltip.test.tsx]: probably you wanted to us `attributes` in the first one

File: packages/tooltip/Tooltip.tsx [hunk 8]
```diff
@@ -9,34 +9,45 @@ const TooltipTypes = {
...
-type TooltipTypesType = typeof TooltipTypes;
-
 // public
-interface TooltipProps {
-  type?: keyof TooltipTypesType,
-  label: string,
-}
+type TooltipProps = {
+  type?: keyof typeof TooltipTypes,
+  label?: string,
```
Comment (Hotell) [inline packages/tooltip/Tooltip.tsx]: label should be required

**Review by Hotell — CHANGES_REQUESTED** [review]

**Review by Hotell — APPROVED** [review]

**Pull request merged**